query/src/main/java/org/infinispan/query/affinity/AffinityIndexManager.java

```diff
@@ -0,0 +1,286 @@
+import org.hibernate.search.spi.SearchIntegrator;
+import org.hibernate.search.spi.WorkerBuildContext;
+import org.hibernate.search.store.DirectoryProvider;
+import org.infinispan.Cache;
+import org.infinispan.distribution.DistributionManager;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.hibernate.search.spi.InfinispanDirectoryProvider;
+import org.infinispan.lucene.impl.DirectoryExtensions;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.TopologyChanged;
+import org.infinispan.notifications.cachelistener.event.TopologyChangedEvent;
+import org.infinispan.query.backend.ComponentRegistryService;
+import org.infinispan.query.backend.KeyTransformationHandler;
+import org.infinispan.query.backend.QueryInterceptor;
+import org.infinispan.query.logging.Log;
+import org.infinispan.remoting.responses.Response;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.util.ByteString;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+
+/**
+ * {@link org.hibernate.search.indexes.spi.IndexManager} that splits the index into shards.
+ *
+ * @author gustavonalle
+ * @since 8.2
+ */
+public class AffinityIndexManager extends DirectoryBasedIndexManager implements OperationFailedHandler {
+
+   private static final Log log = LogFactory.getLog(AffinityIndexManager.class, Log.class);
+   private static final int POLL_WAIT = 100;
+
+   private RpcManager rpcManager;
+   private DistributionManager distributionManager;
+   private KeyTransformationHandler keyTransformationHandler;
+   private Cache cache;
+
+   private volatile boolean hasOwnership = true;
+   private volatile Address nextOwner = null;
+
+   private final ReadWriteLock flushLock = new ReentrantReadWriteLock();
+   private final Lock writeLock = flushLock.writeLock();
+   private final Lock readLock = flushLock.readLock();
+
+
+   @Override
+   public void lockObtainFailed(List<LuceneWork> failingOperations) {
+      logDebug("Operation %s failed due to lock already in used", failingOperations);
+      try {
+         Thread.sleep(POLL_WAIT);
+      } catch (InterruptedException e) {
+         Thread.currentThread().interrupt();
+      }
+      Address lockHolder = getLockHolder(getIndexName());
+      clearIfNeeded(lockHolder);
+      if (failingOperations.stream().anyMatch(w -> w.getIdInString() != null)) {
+         logDebug("Retrying operations %s", failingOperations);
+         CompletableFuture<Void> future = CompletableFuture.supplyAsync(() -> {
```

Comment:
I wondered the same, but since this code path can only be reached sometimes during topology changes, I opted to leave it to the fj pool. On a second though it'd be better to use the ASYNC_OPERATIONS_EXECUTOR  :+1: